pkg/plugins/events/eventgrid/eventgrid.go

```diff
@@ -0,0 +1,209 @@
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/date"
+	"github.com/nitric-dev/membrane/pkg/plugins/errors"
+	"github.com/nitric-dev/membrane/pkg/plugins/errors/codes"
+	"github.com/nitric-dev/membrane/pkg/plugins/events"
+	azureutils "github.com/nitric-dev/membrane/pkg/providers/azure/utils"
+	"github.com/nitric-dev/membrane/pkg/utils"
+)
+
+type EventGridEventService struct {
+	events.UnimplementedeventsPlugin
+	client      eventgridapi.BaseClientAPI
+	topicClient eventgridmgmtapi.TopicsClientAPI
+}
+
+func (s *EventGridEventService) ListTopics() ([]string, error) {
+	newErr := errors.ErrorsWithScope(
+		"EventGrid.ListTopics",
+		map[string]interface{}{
+			"list": "topics",
+		},
+	)
+	//Set the topic page length
+	pageLength := int32(10)
+
+	ctx := context.Background()
+	results, err := s.topicClient.ListBySubscription(ctx, "", &pageLength)
+
+	if err != nil {
+		return nil, newErr(
+			codes.Internal,
+			"error listing by subscription",
+			err,
+		)
+	}
+
+	var topics []string
+
+	//Iterate over the topic pages adding their names to the topics slice
+	for results.NotDone() {
+		topicsList := results.Values()
+		for _, topic := range topicsList {
+			topics = append(topics, *topic.Name)
+		}
+		results.NextWithContext(ctx)
+	}
+
+	return topics, nil
+}
+
+func (s *EventGridEventService) GetTopicEndpoint(topicName string) (string, error) {
+	ctx := context.Background()
+	pageLength := int32(10)
+	results, err := s.topicClient.ListBySubscription(ctx, "", &pageLength)
+	if err != nil {
+		return "", fmt.Errorf(err.Error())
+	}
+
+	for results.NotDone() {
+		topicsList := results.Values()
+		for _, topic := range topicsList {
+			if *topic.Name == topicName {
+				return strings.TrimSuffix(strings.TrimPrefix(*topic.Endpoint, "https://"), "/api/events"), nil
+			}
+		}
+		results.Next()
+	}
+	return "", fmt.Errorf("topic with provided name could not be found")
+}
+
+func (s *EventGridEventService) NitricEventToEvent(topic string, event *events.NitricEvent) ([]eventgrid.Event, error) {
+	payload, err := json.Marshal(event.Payload)
+	if err != nil {
+		return nil, err
+	}
+	dataVersion := "1.0"
+	azureEvent := []eventgrid.Event{
+		{
+			ID:          &event.ID,
+			Data:        &payload,
+			EventType:   &event.PayloadType,
+			Subject:     &topic,
+			EventTime:   &date.Time{time.Now()},
+			DataVersion: &dataVersion,
+		},
+	}
+
+	return azureEvent, nil
+}
+
+func (s *EventGridEventService) Publish(topic string, event *events.NitricEvent) error {
+	newErr := errors.ErrorsWithScope(
+		"EventGrid.Publish",
+		map[string]interface{}{
+			"topic": topic,
+		},
+	)
+	ctx := context.Background()
+
+	if len(topic) == 0 {
+		return newErr(
+			codes.InvalidArgument,
+			"provided invalid topic",
+			fmt.Errorf(""),
+		)
+	}
+	if event == nil {
+		return newErr(
+			codes.InvalidArgument,
+			"provided invalid event",
+			fmt.Errorf(""),
```

Comment:
This can be nil instead of an empty error. Alternatively, you could provide more detail in this "cause" error message. E.g. "non-nil event is required".